### PR TITLE
Pg tests templates

### PR DESCRIPTION
**Pull request by clemens-msupply**
# 👩🏻‍💻 What does this PR do?

- Uses DB template to speed up Postgres tests (similar to what is already done in Sqlite)
- Add option to disable "template tests" for Sqlite and Postgres. Easy to do and might be useful if there is a bug in the template code... (see updated readme)
- For Sqlite the templates are only generated for the repository crate now instead for all crates

Some numbers (from a single runs while keep working on my machine):

- Sqlite no templates: 30m2.075s (this is really long now!)
- Sqlite with template: 30.863s to ~1min (warm)
- Postgres no templates: 11m18
- Postgres with templates: 6m47.066
- Postgres no templates + tmpfs: 1m58.474s
- Postgres with template + tmpfs: 1m2.668

That's how I use a Postgres docker container with a tmpfs:
```
sudo docker run -p 5432:5432 --tmpfs /var/lib/pg/data -e PGDATA=/var/lib/pg/data -e POSTGRES_PASSWORD=password --name pgdev-tmp-database --shm-size=1g --rm postgres -N 500
```

There is some further room for improvments/ TODO:

- Don't generate markers other than for the repository crate
- More mock test data can be cached. Currently, only the full mock dataset is cached and if a test uses a subset it inserts the required mock data. This might result in a further speed up...

# 🧪 Testing

Check that sqlite + postgres test are still working

### CODE REVIEW

File: server/repository/src/test_db/sqlite.rs [hunk 1]
```diff
@@ -38,83 +70,78 @@ pub(crate) async fn setup_with_version(
     inserts: MockDataInserts,
 ) -> (StorageConnectionManager, MockDataCollection) {
     let db_path = db_settings.connection_string();
+    let memory_mode = db_path.starts_with("file:");
+    if memory_mode || env_msupply_no_test_db_template() {
+        return setup_with_version_no_template(db_settings, version, inserts).await;
+    }
```
Comment (clemens-msupply) [inline server/repository/src/test_db/sqlite.rs:76]: This is the main change, changes below are mostly from changing the level of indentation...

File: server/repository/src/test_db/sqlite.rs [hunk 2]
```diff
@@ -38,83 +70,78 @@ pub(crate) async fn setup_with_version(
...
-                    fs::remove_file(&entry.path()).unwrap();
-                }
+        format!(
+            "___template_{}",
+            version.as_ref().unwrap_or(&Version::from_package_json())
+        )
+    };
+
+    let guard = TEMPLATE_LOCK.lock().unwrap();
+    let template_output_dir = template_dir();
```
Comment (clemens-msupply) [inline server/repository/src/test_db/sqlite.rs:93]: This changed though to always use the repository crate to store templates

**Comment by jmbrunskill** [comment]
On M3 mac...

**SQLite**
```
cargo test  21.40s user 41.94s system 223% cpu 28.313 total
MSUPPLY_NO_TEST_DB_TEMPLATE=true cargo test  1120.34s user 6687.71s system 709% cpu 18:21.21 total
```

**Postgres**
```
cargo test --features=postgres  22.04s user 12.81s system 23% cpu 2:30.10 total
MSUPPLY_NO_TEST_DB_TEMPLATE=true cargo test --features=postgres  36.92s user 24.01s system 36% cpu 2:44.95 total
```

**Postgres Tmp File System**
Didn't work for me getting this error...
`called `Result::unwrap()` on an `Err` value: DBError { msg: "UNKNOWN", extra: "\"\\\"could not extend file \\\\\\\"base/308410/312386\\\\\\\": No space left on device\\\"\"" }`
I'm sure it's resolvable but don't want to spend more time on it. Probably the default TMPFS is smaller on docker for mac or something?

**Review by jmbrunskill — APPROVED** [review]
Looks good to me @clemens-msupply 
Not such a big speed up with my Mac but all appears to be working.

Thanks for your efforts on this, I didn't realise how slow the tests on SQLite would be without this!

Might look at the tmpfs thing another day...

**Comment by clemens-msupply** [comment]
> On M3 mac...
> 
> **SQLite**
> 
> ```
> cargo test  21.40s user 41.94s system 223% cpu 28.313 total
> MSUPPLY_NO_TEST_DB_TEMPLATE=true cargo test  1120.34s user 6687.71s system 709% cpu 18:21.21 total
> ```
> 
> **Postgres**
> 
> ```
> cargo test --features=postgres  22.04s user 12.81s system 23% cpu 2:30.10 total
> MSUPPLY_NO_TEST_DB_TEMPLATE=true cargo test --features=postgres  36.92s user 24.01s system 36% cpu 2:44.95 total
> ```
> 

Thanks for running the test, this looks consistent with what I tested on my machine. Looks like the migrations on Sqlite are very slow...

**Pull request merged**